Source/WebCore/css/CSSSelector.cpp

```diff
@@ -979,23 +979,28 @@ CSSSelector::CSSSelector(const CSSSelector& other)
     }
 }
 
-void CSSSelector::visitAllSimpleSelectors(auto& apply) const
+bool CSSSelector::visitAllSimpleSelectors(auto& apply) const
 {
-    // Effective C++ advices for this cast to deal with generic const/non-const member function.
-    apply(*const_cast<CSSSelector*>(this));
-
-    // Visit the selector list member (if any) recursively (such as: :has(<list>), :is(<list>),...)
-    if (auto selectorList = this->selectorList()) {
-        auto next = selectorList->first();
-        while (next) {
-            next->visitAllSimpleSelectors(apply);
-            next = CSSSelectorList::next(next);
+    auto current = this;
+    while (current) {
+        // Effective C++ advices for this cast to deal with generic const/non-const member function.
+        if (apply(*const_cast<CSSSelector*>(current)))
+            return true;
+
+        // Visit the selector list member (if any) recursively (such as: :has(<list>), :is(<list>),...)
+        if (auto selectorList = current->selectorList()) {
+            auto next = selectorList->first();
+            while (next) {
+                if (next->visitAllSimpleSelectors(apply))
```

Comment:
Should this be converted to iteration too?

Reply:
Ideally yes, but it's more complex (need a queue?) and not sure it's worth it. Maybe for another PR 😅

Reply:
Fixed ✅